exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,618 @@
+    timer.expires_after(std::chrono::milliseconds(20));
+
+    timer.async_wait([&](boost::system::error_code error_code) {
+      if (!error_code) {
+        cancel_signal_.emit(asio::cancellation_type::all);
+        read_finished = true;
+      }
+    });
+
+    asio::co_spawn(mqtt_client_->strand(), async_receive_routine(receiver, read_finished, signal_value),
+                   asio::bind_cancellation_slot(cancel_signal_.slot(), boost::asio::detached));
+
+    while (!read_finished) {
+      co_await asio::steady_timer{ io_ctx_, std::chrono::milliseconds{ 1 } }.async_wait(asio::use_awaitable);
+    }
+
+    if (signal_value != std::nullopt) {
+      set_value_payload(variable_metric, signal_value.value());
+    } else {
+      variable_metric->set_is_null(true);
+    }
+  }
+
+  auto async_receive_routine(auto&& receiver, bool& read_finished, std::optional<std::any>& signal_value_)
+      -> asio::awaitable<void> {
+    auto val = co_await receiver->async_receive(asio::use_awaitable);
+    if (val.has_value()) {
+      signal_value_ = val.value();
+    }
+    read_finished = true;
+  }
+
+  auto timer(std::chrono::steady_clock::duration dur) -> asio::awaitable<void> {
+    asio::steady_timer timer(co_await asio::this_coro::executor);
+    timer.expires_after(dur);
+    co_await timer.async_wait(asio::use_awaitable);
+    cancel_signal_.emit(asio::cancellation_type::total);
+  }
+
+  auto send_message(std::string topic, std::string payload, async_mqtt::qos _qos) -> asio::awaitable<void> {
+    // flag is needed because it is not possible to co_await inside a try/catch block
+    bool send_failed = false;
+    try {
+      uint16_t packet_id = 0;
+      if (_qos != async_mqtt::qos::at_most_once) {
+        packet_id = mqtt_client_->acquire_unique_packet_id().value();
+      }
+      logger_.info("Attempting to send message to topic: {}", topic);
+      auto result =
+          co_await mqtt_client_->send(async_mqtt::v5::publish_packet{ packet_id, async_mqtt::allocate_buffer(topic),
+                                                                      async_mqtt::allocate_buffer(payload), _qos },
+                                      asio::use_awaitable);
+
+      if (result) {
+        logger_.error("Failed to send message to topic: {}", topic);
+        send_failed = true;
+      } else {
+        logger_.info("Message successfully sent to topic: {}", topic);
+      }
+    } catch (std::exception const& exc) {
+      logger_.error("Exception occurred while sending message to topic: {}. Exception: {}", topic, exc.what());
+      send_failed = true;
+    }
+
+    if (send_failed) {
+      logger_.info("Attempting to reconnect to broker");
+      co_await asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::use_awaitable);
+      logger_.info("Reconnected to broker. Attempting to resend the message to topic: {}", topic);
+      co_await send_message(topic, payload, _qos);
+    }
+  }
```

Comment:
Can you not just use the catch?
```suggestion

    try {
      uint16_t packet_id = 0;
      if (_qos != async_mqtt::qos::at_most_once) {
        packet_id = mqtt_client_->acquire_unique_packet_id().value();
      }
      logger_.info("Attempting to send message to topic: {}", topic);
      auto result =
          co_await mqtt_client_->send(async_mqtt::v5::publish_packet{ packet_id, async_mqtt::allocate_buffer(topic),
                                                                      async_mqtt::allocate_buffer(payload), _qos },
                                      asio::use_awaitable);

      if (result) {
        logger_.error("Failed to send message to topic: {}", topic);
        send_failed = true;
      } else {
        logger_.info("Message successfully sent to topic: {}", topic);
      }
    } catch (std::exception const& exc) {
      logger_.error("Exception occurred while sending message to topic: {}. Exception: {}", topic, exc.what());
      logger_.info("Attempting to reconnect to broker");
      co_await asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::use_awaitable);
      logger_.info("Reconnected to broker. Attempting to resend the message to topic: {}", topic);
      co_await send_message(topic, payload, _qos);
    }
  }
```

Reply:
If the result from the send is not successful then a reconnect won't happen.

```cpp
if (result) {
    logger_.error("Failed to send message to topic: {}", topic);
    send_failed = true;
}
```

this doesn't result in a reconnect.

Reply:
after reading the send function in async_mqtt I don't think it throws an exception. This should be enough:

```cpp
auto send_message(std::string topic, std::string payload, async_mqtt::qos _qos) -> asio::awaitable<void> {
    uint16_t packet_id = 0;
    if (_qos != async_mqtt::qos::at_most_once) {
        packet_id = mqtt_client_->acquire_unique_packet_id().value();
    }
    
    logger_.info("Attempting to send message to topic: {}", topic);

    auto error_code = co_await mqtt_client_->send(async_mqtt::v5::publish_packet{ packet_id, async_mqtt::allocate_buffer(topic), async_mqtt::allocate_buffer(payload), _qos }, use_nothrow_awaitable);

    if (error_code) {
        logger_.error("Failed to send message to topic: {}", topic);
        logger_.info("Attempting to reconnect to broker");
      
        co_await asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::use_awaitable);
        logger_.info("Reconnected to broker. Attempting to resend the message to topic: {}", topic);
        co_await send_message(topic, payload, _qos);
    } else {
        logger_.info("Message successfully sent to topic: {}", topic);
    }
}
```